### PR TITLE
feat: added pseudo labeler script

### PR DESCRIPTION
**Pull request by hasibzunair**
Example code snippet to run the new feature: 

```
import decavision

pseudolabels = decavision.dataset_preparation.generate_pseudolabels.PseudoLabelGenerator()
pseudolabels.generate_pseudolabel_data(model_path="logs/yogapose_teacher/yogapose_teacher.h5",
                                    train_data_path="datasets/yogapose/train",
                                    unlabeled_path="datasets/yogapose/unlabeled",
                                    pseudo_data_path="datasets/yogapose/train_ssl",
                                    save_confidences_csv=True,
                                    plot_confidences=True,
                                    threshold=None)
```

### CODE REVIEW

File: decavision/dataset_preparation/generate_pseudolabels.py [hunk 1]
```diff
@@ -0,0 +1,159 @@
+import os
+import pandas as pd
+import tensorflow as tf
```
Comment (yangobeil) [inline decavision/dataset_preparation/generate_pseudolabels.py]: In the imports the convention that I like to use is to have in general 3 blocks separated by empty lines. The first block is importing the packages that are included in basic python (like os). The second block is for installed packages (like pandas). The third block is for when you import code from other scripts that you wrote. In all of the blocks I think it's also nice to keep the imports in alphabetical order.

File: decavision/dataset_preparation/generate_pseudolabels.py [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+            os.mkdir("outputs")
+        if save_confidences_csv:
+            print("Saving CSV with pseudo predictions.")
+            data = {'filepaths': raw_predictions_paths,
+                    'class preds highest': raw_predictions,
+                    'class preds all': raw_predictions_all}
+            df = pd.DataFrame(data)
+            df.to_csv("outputs/pseudo_predictions.csv", index=False)
+
+        if plot_confidences:
```
Comment (yangobeil) [inline decavision/dataset_preparation/generate_pseudolabels.py:234]: I would put the code to make the plots into a separate method. There are a few reasons why. First you seem to be repeating a lot of lines with the two types of plots so having a method to use would avoid that. Second people could plot the data as they like by calling the method instead of having to recalculate the pseudolabels every time. Finally this method is pretty big so splitting the logically distinct parts is a good thing to do.

File: decavision/dataset_preparation/generate_pseudolabels.py [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+
+        print("Generating pseudo labels...")
+        # Generate pseudo labels
+        unlabeled_image_paths = os.listdir(unlabeled_path)
+        print("There are {} unlabeled images.".format(
+            len(unlabeled_image_paths)))
+        raw_predictions_paths = []
+        raw_predictions = []  # single confidence value of predicted class
+        raw_predictions_all = []  # confidences for all classes
+        for path in tqdm(unlabeled_image_paths[:100]):
```
Comment (yangobeil) [inline decavision/dataset_preparation/generate_pseudolabels.py]: Like I mentionned before it would be much faster to make the predictions in batches.

File: decavision/dataset_preparation/generate_pseudolabels.py [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+        # Resize
+        image = tf.image.resize(image, size=[*target_size])
+        image = tf.reshape(image, [*target_size, 3])
+        image = tf.cast(image, tf.uint8)
+        # Normalize [0, 1]
+        image = tf.image.convert_image_dtype(image, dtype=tf.float32)
+        image = image.numpy()
+        return image
+
+    def generate_pseudolabel_data(self, model_path,
```
Comment (yangobeil) [inline decavision/dataset_preparation/generate_pseudolabels.py]: Some of these arguments should be inputs to the class instead of this method I think. For example the paths because they could be used in other methods.

File: decavision/dataset_preparation/generate_pseudolabels.py [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+            # Get class index
+            y = np.argmax(y_pred)
+            # Get probability score
+            proba = y_pred[0][y]
+
+            raw_predictions.append(proba)
+            raw_predictions_all.append(y_pred)
+            raw_predictions_paths.append(os.path.join(unlabeled_path, path))
+
+            # The results variable should be the same as the class category
```
Comment (yangobeil) [inline decavision/dataset_preparation/generate_pseudolabels.py]: I would also split the moving of the images to a different method, again to avoid having to generate the pseudolabels every time you want to do something. I think the method to generate the pseudolabels could save the probabilities in a csv and then the method to move the images uses that csv to decide where to move the images.

File: decavision/dataset_preparation/generate_pseudolabels.py [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+        """
+        # Copy the training/labeled data to the destination folder where
+        # we will also store the pseudo labels.
+        copy_tree(train_data_path, pseudo_data_path)
+
+        # Load model
+        model = load_model(model_path, compile=False)
+        print("Loaded model.")
+
+        # Make dictionary for classes and their index
```
Comment (yangobeil) [inline decavision/dataset_preparation/generate_pseudolabels.py]: With the format that I am aiming to use for the different methods I think this would fit better in the __init__.

File: decavision/dataset_preparation/generate_pseudolabels.py (hunk 5 above)
Comment (yangobeil) [reply decavision/dataset_preparation/generate_pseudolabels.py]: The csv can also be used by the methods that generate plots.

File: decavision/dataset_preparation/generate_pseudolabels.py (hunk 2 above)
Comment (hasibzunair) [reply decavision/dataset_preparation/generate_pseudolabels.py:234]: Done.

File: decavision/dataset_preparation/generate_pseudolabels.py (hunk 3 above)
Comment (hasibzunair) [reply decavision/dataset_preparation/generate_pseudolabels.py]: Did not address this as I am not sure how to do it. Since we are not using a dataset but rather a folder of images, we cannot use generators. Writing a custom function would result in more loops and not sure if it would be that fast. Need your suggestions on this.

File: decavision/dataset_preparation/generate_pseudolabels.py (hunk 4 above)
Comment (hasibzunair) [reply decavision/dataset_preparation/generate_pseudolabels.py]: Added all path arguments in `init`.

File: decavision/dataset_preparation/generate_pseudolabels.py (hunk 5 above)
Comment (hasibzunair) [reply decavision/dataset_preparation/generate_pseudolabels.py]: Made a separate method.

File: decavision/dataset_preparation/generate_pseudolabels.py (hunk 6 above)
Comment (hasibzunair) [reply decavision/dataset_preparation/generate_pseudolabels.py]: Done!

File: decavision/dataset_preparation/generate_pseudolabels.py [hunk 7]
```diff
@@ -0,0 +1,220 @@
...
+            # Plot only if there are any CSV files.
+            if save_confidences_csv:
+                print("Plotting data.")
+                self._plot_confidence_scores(
+                    csv_path="outputs/data.csv", classes=class_names)
+            else:
+                print("No CSV file is present to plot the data.")
+
+
+if __name__ == '__main__':
```
Comment (yangobeil) [inline decavision/dataset_preparation/generate_pseudolabels.py]: This is not necessary. The code in this script is made to be imported and not run from the console.

Comment (hasibzunair) [reply decavision/dataset_preparation/generate_pseudolabels.py]: Done!

File: decavision/dataset_preparation/generate_pseudolabels.py (hunk 3 above)
Comment (yangobeil) [reply decavision/dataset_preparation/generate_pseudolabels.py]: The idea would be to feed the whole list of image paths to tf.data and generate a tf.dataset object from it with batches of images that you would feed into the model to predict. If you are not sure what I mean we can discuss it in our meeting tomorrow :)

Comment (hasibzunair) [reply decavision/dataset_preparation/generate_pseudolabels.py]: Okay, we can discuss tomorrow!

File: decavision/dataset_preparation/generate_pseudolabels.py [hunk 8]
```diff
@@ -0,0 +1,216 @@
...
+
+    def __init__(self, model_path, train_data_path, unlabeled_path, pseudo_data_path):
+        self.model_path = model_path
+        self.train_data_path = train_data_path
+        self.unlabeled_path = unlabeled_path
+        self.pseudo_data_path = pseudo_data_path
+
+        # Load model
+        model = load_model(self.model_path, compile=False)
+        self.model = model
```
Comment (yangobeil) [inline decavision/dataset_preparation/generate_pseudolabels.py]: You can directly set self.model to be load.model.

File: decavision/dataset_preparation/generate_pseudolabels.py [hunk 9]
```diff
@@ -0,0 +1,216 @@
...
+            title = "Confidences for the class: {}".format(classes[idx])
+            path = "outputs/{}_confidences.png".format(classes[idx])
+            self._plot_data(predictions, title, path)
+
+        # Plot graph for highest confidence pseudo labels for all unlabeled images
+        self._plot_data(raw_predictions,
+                        name="Highest confidence pseudo labels",
+                        output_path="outputs/highest_confidence_predictions.png")
+
+    def _save_pseudo_labels(self, csv_path, threshold, dictionary):
```
Comment (yangobeil) [inline decavision/dataset_preparation/generate_pseudolabels.py]: I would change the name of this method. It's not clear what it does to maybe something like move_unlabeled_images would be clearer.

Comment (yangobeil) [reply decavision/dataset_preparation/generate_pseudolabels.py]: Also I would make this method public (remove the _ at the beginning) along with _plot_confidence_scores to make it available for people to use.

File: decavision/dataset_preparation/generate_pseudolabels.py [hunk 10]
```diff
@@ -0,0 +1,216 @@
...
+            print("Saving CSV with pseudo predictions.")
+            data = {'Filepaths': raw_predictions_paths,
+                    'Predicted Class': predicted_class,
+                    'Highest Confidence': raw_predictions,
+                    'All Class Predictions': raw_predictions_all}
+            df = pd.DataFrame(data)
+            df.to_csv("outputs/data.csv", index=False)
+
+        # Save pseudo labeled images
+        self._save_pseudo_labels(csv_path="outputs/data.csv",
```
Comment (yangobeil) [inline decavision/dataset_preparation/generate_pseudolabels.py]: Won't you have a problem if the user decides to not use a csv? I would remove the option to not use a csv and maybe add the csv path to the __init__ method to avoid repeating it.

File: decavision/dataset_preparation/generate_pseudolabels.py [hunk 11]
```diff
@@ -0,0 +1,216 @@
...
+    """
+    Class to generate pseudo labels.
+    Arguments:
+        model_path (str): path to trained model
+        train_path (str): path to training data
+        unlabeled_path (str): path to unlabeled data
+        pseudo_path (str): path to store train data and pseudo data
+    """
+
+    def __init__(self, model_path, train_data_path, unlabeled_path, pseudo_data_path):
```
Comment (yangobeil) [inline decavision/dataset_preparation/generate_pseudolabels.py]: It would be nice to have generic default values for the patsh so people can use the class easily by naming their folders correctly. Make sure to follow the convention used in the rest of the library.

File: decavision/dataset_preparation/generate_pseudolabels.py (hunk 10 above)
Comment (yangobeil) [reply decavision/dataset_preparation/generate_pseudolabels.py]: I would add an argument to decide if this step (moving the images) is done or not. This way it would be possible to generate the csv without automatically move the images.

File: decavision/dataset_preparation/generate_pseudolabels.py [hunk 12]
```diff
@@ -0,0 +1,216 @@
...
+                pseudo_data_path: A folder with both labeled and pseudo labeled images.
+        """
+
+        # Copy the training/labeled data to the destination folder where
+        # we will also store the pseudo labels.
+        copy_tree(self.train_data_path, self.pseudo_data_path)
+
+        # Make dictionary for classes and their index
+        class_names = sorted(os.listdir(self.train_data_path))
+        class_dict = {}
```
Comment (yangobeil) [inline decavision/dataset_preparation/generate_pseudolabels.py]: Here are two suggestions to make this code much cleaner:
- use enumerate instead of making your own index.
- use list/dict comprehension.

The result will look like
```
class_dict = {cat, i for i, cat in enumerate(class_names)}
```

File: decavision/dataset_preparation/generate_pseudolabels.py [hunk 13]
```diff
@@ -0,0 +1,216 @@
...
+                plot_confidences (boolean): Whether to plot confidence graphs for raw confidences and per class confidences.
+                threshold (float): Discard images with prediction below this confidence, default is None.
+
+            Returns:
+                pseudo_data_path: A folder with both labeled and pseudo labeled images.
+        """
+
+        # Copy the training/labeled data to the destination folder where
+        # we will also store the pseudo labels.
+        copy_tree(self.train_data_path, self.pseudo_data_path)
```
Comment (yangobeil) [inline decavision/dataset_preparation/generate_pseudolabels.py]: This should be in the _save_pseudo_labels method to have all the steps involving movement of images at the same place.

File: decavision/dataset_preparation/generate_pseudolabels.py [hunk 14]
```diff
@@ -0,0 +1,216 @@
...
+        Arguments:
+            csv_path (str): Path to save SCV file
+            classes (list): List containing the class names
+        """
+        dt = pd.read_csv(csv_path)
+        dt['All Class Predictions List'] = dt['All Class Predictions'].apply(
+            lambda x: ast.literal_eval(x))
+
+        raw_predictions_ = dt[["Highest Confidence"]].values
+        raw_predictions = [raw_predictions_[idx][0]
```
Comment (yangobeil) [inline decavision/dataset_preparation/generate_pseudolabels.py]: Maybe this could be done more clearly with the apply method of pandas and a lambda function. In any case, if I understand your code correctly, this list comprehension could be written more simply the following way:
```
[pred[0] for pred in raw_predictions_]

File: decavision/dataset_preparation/generate_pseudolabels.py [hunk 15]
```diff
@@ -0,0 +1,216 @@
...
+                           for idx in range(len(raw_predictions_))]
+
+        raw_predictions_all_ = dt[['All Class Predictions List']].values
```
Comment (yangobeil) [inline decavision/dataset_preparation/generate_pseudolabels.py:106]: Similar comment to the one just above.

File: decavision/dataset_preparation/generate_pseudolabels.py [hunk 16]
```diff
@@ -0,0 +1,216 @@
...
+        raw_predictions_all = [raw_predictions_all_[idx][0]
+                               for idx in range(len(raw_predictions_all_))]
+
+        # Plot graph for highest confidence pseudo labels for each class
+        for idx in range(len(classes)):
```
Comment (yangobeil) [inline decavision/dataset_preparation/generate_pseudolabels.py]: A few suggestions again:
- It could be cleaner to use enumerate instead of range(len(...))
- You can use a list comprehension to define 'predictions'

File: decavision/dataset_preparation/generate_pseudolabels.py [hunk 17]
```diff
@@ -0,0 +1,216 @@
...
+        """
+        Plots a bar plot.
+
+        Arguments:
+            predictions (list): List of predictions
+            name (str): Title of the plot
+            output_path (str): Save file using this name
+        """
+        predictions = sorted(predictions)
+        samples = [pred for pred in range(len(predictions))]
```
Comment (yangobeil) [inline decavision/dataset_preparation/generate_pseudolabels.py]: This line could be more cleanly written as list(range(len(predictions))).

File: decavision/dataset_preparation/generate_pseudolabels.py (hunk 8 above)
Comment (hasibzunair) [reply decavision/dataset_preparation/generate_pseudolabels.py]: Done.

File: decavision/dataset_preparation/generate_pseudolabels.py (hunk 9 above)
Comment (hasibzunair) [reply decavision/dataset_preparation/generate_pseudolabels.py]: Done.

File: decavision/dataset_preparation/generate_pseudolabels.py (hunk 10 above)
Comment (hasibzunair) [reply decavision/dataset_preparation/generate_pseudolabels.py]: Done.

File: decavision/dataset_preparation/generate_pseudolabels.py (hunk 12 above)
Comment (hasibzunair) [reply decavision/dataset_preparation/generate_pseudolabels.py]: Done.

File: decavision/dataset_preparation/generate_pseudolabels.py (hunk 13 above)
Comment (hasibzunair) [reply decavision/dataset_preparation/generate_pseudolabels.py]: Done.

File: decavision/dataset_preparation/generate_pseudolabels.py (hunk 14 above)
Comment (hasibzunair) [reply decavision/dataset_preparation/generate_pseudolabels.py]: Done.

File: decavision/dataset_preparation/generate_pseudolabels.py (hunk 15 above)
Comment (hasibzunair) [reply decavision/dataset_preparation/generate_pseudolabels.py:106]: Done.

File: decavision/dataset_preparation/generate_pseudolabels.py (hunk 16 above)
Comment (hasibzunair) [reply decavision/dataset_preparation/generate_pseudolabels.py]: Done.

File: decavision/dataset_preparation/generate_pseudolabels.py (hunk 11 above)
Comment (hasibzunair) [reply decavision/dataset_preparation/generate_pseudolabels.py]: Done.

File: decavision/dataset_preparation/generate_pseudolabels.py (hunk 17 above)
Comment (hasibzunair) [reply decavision/dataset_preparation/generate_pseudolabels.py]: Done.

File: requirements.txt [hunk 18]
```diff
@@ -11,4 +11,5 @@ pyunpack==0.2.2
 scikit-learn==0.23.2
 scikit-optimize==0.8.1
 seaborn==0.11.0
-tensorflow==2.3.1
\ No newline at end of file
+tensorflow==2.4.0
```
Comment (yangobeil) [inline requirements.txt:14]: I am not sure why you changed the tensorflow version. Did you find a bug with 2.3.1?

**Comment by yangobeil** [comment]
I made a new commit with a few minor changes to the code just to make it a bit more user friendly.

Apart from the batches I think the only thing left to do would be to add a method to combine the train_ssl and train folders to make a final dataset.

File: requirements.txt (hunk 18 above)
Comment (hasibzunair) [reply requirements.txt:14]: Checked it. The warning is due to TF 2.3, we should use 2.4, the default one of Colab.

**Comment by hasibzunair** [comment]
> Apart from the batches I think the only thing left to do would be to add a method to combine the train_ssl and train folders to make a final dataset.

The folder `train_ssl` already has the training dataset. `train_ssl` is basically training data and pseudo labels combined. I made a copy of the training dataset in a new folder `train_ssl` using `copy_tree(self.train_data_path, self.pseudo_data_path)` prior to adding pseudo labels in `train_ssl`. Did not save the pseudo labels separately as I didn't think there is any use for it. The CSV file already has them.

**Comment by yangobeil** [comment]
I tested the script and it works nicely! The only detail that I would change is iterate over the batches one at a time when making predictions. This can be done with something like
```
for batch in ds:
  y_preds = model.predict(batch)
```
I think this is slower but it will allow you to use tqdm to keep track of progress. It will be useful because the prediction takes a lot of time with many images so it's important to have an idea of where it is at.

**Review by yangobeil — APPROVED** [review]
Every looks good now. Very nice job!

**Comment by hasibzunair** [comment]
Yay! Thanks for reviews, was really helpful! :D
